### DISCUSSION TITLE
0.12.0

### DISCUSSION THREAD

**Discussion by MultisampledNight** [neovide/neovide#2197] [General]
### Breaking

* fix!: Improve render loop by @fredizzimo in https://github.com/neovide/neovide/pull/1977
  * Command line options starting with `no` have been renamed and now start with `--no-`
    * All options are kebab-case now
  * Multigrid has been made default
  * The transparency rendering on macOS has been changed, and a full fix is work in progress here https://github.com/neovide/neovide/pull/2168. Meanwhile, it's recommended to unset `neovide_background_color` and set `neovide_transparency` according to your preference. Note that the the title bar will remain solid with the default OS colors. 
  * NOTE: This is a big change, and a lot of issues have been fixed, check #1977 for the full list.
* feat!: Add underline scale factor config by @Kethku in https://github.com/neovide/neovide/pull/2116
  * `underline_automatic_scaling` has been removed
* feat!: use display link api to implement vsync on macos. by @crupest in https://github.com/neovide/neovide/pull/2102

### Features

* Added numpad support by @plague-spreader in https://github.com/neovide/neovide/pull/2032
* Tell neovim focus was gained/lost via 'ui_set_focus' by @9mm in https://github.com/neovide/neovide/pull/2033
* feat: minimize window on suspend by @SyedAhkam in https://github.com/neovide/neovide/pull/1971
* feat: Better reporting of unexpected output from neovim by @fredizzimo in https://github.com/neovide/neovide/pull/2083
* feat: add metadata to Cargo.toml in prep for crates.io publish by @Kethku in https://github.com/neovide/neovide/pull/2103
* Add drop shadow to floating windows by @Kethku in https://github.com/neovide/neovide/pull/2028
* feat: persist grid size along with window size by @sid-6581 in https://github.com/neovide/neovide/pull/2127
* Extra context if can't save to file by @dev-ardi in https://github.com/neovide/neovide/pull/2133
* fix!: Ignore window size when minimized by @fredizzimo in https://github.com/neovide/neovide/pull/2153
* feat: NEOVIDE_FORK environment variable by @9mm in https://github.com/neovide/neovide/pull/2189
* feat: X1 and X2 mouse button support by @agraven in https://github.com/neovide/neovide/pull/2161
* feat: Smoother render loop by @fredizzimo in https://github.com/neovide/neovide/pull/2188

### Fixes

* fix: fix horizontal scroll by @yasen-pavlov in https://github.com/neovide/neovide/pull/2021
* fix: Fix some mappings involving shift by @fredizzimo in https://github.com/neovide/neovide/pull/2018
* Fix config change notification using remote neovim by @mopp in https://github.com/neovide/neovide/pull/2067
* set neovide_channel_id always by @mopp in https://github.com/neovide/neovide/pull/2070
* Tokenize command arguments in `$NEOVIM_BIN` (fix #2060) by @rhysd in https://github.com/neovide/neovide/pull/2063
* fix: run :edit command with nvim_cmd to escape special characters by @ZyX-II in https://github.com/neovide/neovide/pull/2056
* fix: Fix mouse dragging by @fredizzimo in https://github.com/neovide/neovide/pull/2082
* Report error when setting a wrong guifont value by @Tmpod in https://github.com/neovide/neovide/pull/2089
* fix: do not spawn particles when jumping to a location instantly by @Kethku in https://github.com/neovide/neovide/pull/2106
* Observe mousemoveevent option by @Kethku in https://github.com/neovide/neovide/pull/2111
* fix: Fix macOS startup issues by @fredizzimo in https://github.com/neovide/neovide/pull/2107
* use grid 0 to report mouse events when not using multigrid by @Kethku in https://github.com/neovide/neovide/pull/2115
* Fix stacked floating window shadows by @Kethku in https://github.com/neovide/neovide/pull/2114
* Fix remember window size by @Kethku in https://github.com/neovide/neovide/pull/2120
* fix: fix console window not hidden for Windows shell command by @sid-6581 in https://github.com/neovide/neovide/pull/2125
* fix: prevent empty window title by @raggi in https://github.com/neovide/neovide/pull/2126
* Wayland clipboard by @cshuaimin in https://github.com/neovide/neovide/pull/2123
* fix: event loop shouldn't join the render thread by @sid-6581 in https://github.com/neovide/neovide/pull/2129
* fix: Use own data directory instead of nvim's by @agraven in https://github.com/neovide/neovide/pull/2135
* fix: Update winit to 0.29.4 by @fredizzimo in https://github.com/neovide/neovide/pull/2146
* fix: Don't poll the event loop on Wayland to reduce the power usage by @fredizzimo in https://github.com/neovide/neovide/pull/2149
* fix: crash with zero height windows by @fredizzimo in https://github.com/neovide/neovide/pull/2150
* fix: Change the order of vsync wait and swap buffers to reduce latency by @fredizzimo in https://github.com/neovide/neovide/pull/2152
* fix: Ignore negative cursor positions by @fredizzimo in https://github.com/neovide/neovide/pull/2157
* Revert unintentional change to default ime setting by @Kethku in https://github.com/neovide/neovide/pull/2159
* fix: Completely event driven update loop by @fredizzimo in https://github.com/neovide/neovide/pull/2167
* fix: Scrolling of windows with winbar or native borders by @fredizzimo in https://github.com/neovide/neovide/pull/2165

### Docs

* Update troubleshooting.md by @heartbeast42 in https://github.com/neovide/neovide/pull/2026
* Add faq about x11 dpi sizing by @Kethku in https://github.com/neovide/neovide/pull/2105

### Work behind the scenes

* chore: Fix new clippy warnings from rust 1.72 by @fredizzimo in https://github.com/neovide/neovide/pull/2034
* chore!: Update the macOS CI to use macOS 12 instead of 11 by @fredizzimo in https://github.com/neovide/neovide/pull/2108
* Remove unused dependencies by @Kethku in https://github.com/neovide/neovide/pull/2110
* refactor: fix compilation warnings by @sid-6581 in https://github.com/neovide/neovide/pull/2119
* style: fix typos by @MultisampledNight in https://github.com/neovide/neovide/pull/2156
* chore: Improved profiling by @fredizzimo in https://github.com/neovide/neovide/pull/2186

## New Contributors

Thank you all, and also all others who contributed! :heart:

* @heartbeast42 made their first contribution in https://github.com/neovide/neovide/pull/2026
* @yasen-pavlov made their first contribution in https://github.com/neovide/neovide/pull/2021
* @plague-spreader made their first contribution in https://github.com/neovide/neovide/pull/2032
* @9mm made their first contribution in https://github.com/neovide/neovide/pull/2033
* @mopp made their first contribution in https://github.com/neovide/neovide/pull/2067
* @rhysd made their first contribution in https://github.com/neovide/neovide/pull/2063
* @ZyX-II made their first contribution in https://github.com/neovide/neovide/pull/2056
* @SyedAhkam made their first contribution in https://github.com/neovide/neovide/pull/1971
* @Tmpod made their first contribution in https://github.com/neovide/neovide/pull/2089
* @sid-6581 made their first contribution in https://github.com/neovide/neovide/pull/2119
* @raggi made their first contribution in https://github.com/neovide/neovide/pull/2126
* @dev-ardi made their first contribution in https://github.com/neovide/neovide/pull/2133
* @agraven made their first contribution in https://github.com/neovide/neovide/pull/2135
* @crupest made their first contribution in https://github.com/neovide/neovide/pull/2102

**Full Changelog**: https://github.com/neovide/neovide/compare/0.11.2...0.12.0

<hr /><em>This discussion was created from the release <a href='https://github.com/neovide/neovide/releases/tag/0.12.0'>0.12.0</a>.</em>

**Comment by ikimmit** [comment]
Sounds like a great update. Really enjoy the app and its design principles. Thank you for the hard work!

**Comment by agraven** [comment]
Huge shout out to @fredizzimo for all the ongoing optimization and refactoring work, it's really paying off.

**Reply by Kethku** [reply]
I second this! Fredizzimo's work in this release started nearly a year ago. Huge props

**Comment by crupest** [comment]
Thanks for everyone who has made their contribution to bring us such an amazing app. And also thank who have helped me make my first contribution.
I want to give special thanks to @fredizzimo, personally. This is the first time for me to make contribution to such a big community. Without him, I can't make even a little of this. From guiding me to debug Neovide, to helping me learn new techniques and write better code, to finally merging my pr, he has payed so much patience on me, which I can't find an accurate word to describe. He not only cares what code I wrote, but also cares what emotion I had. Even on Christmas night, he was still replying to my questions. (Sorry that we don’t celebrate Christmas in our culture and didn’t notice it until he mentioned it.)
I always believe, the code is dead, but the people behind it are live and full of love. He is exactly the one who perfectly reflects this.
♥️

**Reply by crupest** [reply]
And also @9mm, thank you for testing my prs patiently. I really enjoy working with you!

**Reply by 9mm** [reply]
Thanks @crupest. @fredizzimo is definitely the OG...  just adding another thanks on to the pile for him... Over the last few months he helped fix several issues that were preventing me from migrating my 10 year vim config which I have worked on obsessively. I remember I first tried Neovide and didn't give it too much time, and thought "ah well, too many muscle breaking changes." But I made a few issues and he helped actually sort them all out. Before I knew it I was making first ever commits to Neovide as well. Neovide is really truly great, so thanks to @Kethku for starting it. I prefer it to MacVim which is saying a lot, despite all of MacVims special OSX-centric behavior, I think Neovide feels modern+fresh, and great team behind it is doing awesome. Whoever did all those animations truly has their priority straight 🤣 That would be the very first thing I work on as well if I had a GUI.

**Comment by fredizzimo** [comment]
Thank you very much for the very kind comments here, this is indeed something I have been working on for a long time, and mostly finished during the summer, with some more minor fixes added after that. Most of the process has been frustration that I can't get Windows to work as well as it should, it seems like it's impossible with OpenGL, but I was just to stubborn to give up. We should quite easily be able  to add D3D support ,which should fix that though. 

There are also some explainable stuttering on some wayland gnome systems, that should be investigated a lot deeper, and perhaps together with the Gnome team. 

Also let's not forget that we had **14 new contributors** here. Including @crupest, who has been tackling two very challenging issues, for which we have been looking for contributors since the summer. All the regular neovide maintainers on macOS have been busy with other things.

**Comment by fredizzimo** [comment]
I also want to thank @kchibisov, from the Winit team, who has helped us a lot with debugging of a few issues and putting all this together.

**Comment by bnwa** [comment]
This fixed a bug that was killing me, thank you!

**Comment by sid-6581** [comment]
Awesome work! I had to not use neovide for a couple of years because it had some issues that made it hard to use as a daily driver, but now it's all I use. I appreciate taking my PRs that fix the issues I ran into, looking forward to contributing more!

**Comment by Tmpod** [comment]
I know I come a bit late to the party, but I still wanted to
congratulate everyone involved, and especially the maintainers for such
a wonderful job.

Thank you and have a happy new year! :3

**Comment by bodograumann** [comment]
Sorry to say, but all these smooth-scrollings/window-animations are simply unbearable for me. (Most notably, when auto-scrolling happens from 0,0 to some initial position, e.g. with fugitive's `:Git blame`).
So I imediately disabled them.
Now I am just worried, that the `--no-multigrid` flag might be removed in the future.
What are the plans there?

**Reply by 9mm** [reply]
If you don't already have these they might help. The scroll animation length is probably what you're referring to.

After that, I think you might want multigrid on (if you're only using it to stop animation) as it has some other improvements if I remember right.

```
  -- animation
  vim.g.neovide_cursor_animation_length = 0
  vim.g.neovide_cursor_trail_size = 0
  vim.g.neovide_scroll_animation_length = 0
  vim.g.neovide_cursor_animate_command_line = false
```

**Reply by fredizzimo** [reply]
You could also open specific issue for the initial scroll of git blame, it's not done on purpose, just something that happens due to the implementation, and might be fixable.

And the same thing for you @9mm, I'm not clear which effect you don't like, we might be able to add a configuration to fix that as well.

For me, the smooth scrolling specifically is the best feature of Neovide, as it allows me to track the content during scrolling without loosing focus, so I would obviously hope that we can make it usable for as many people as possible. I also suggest that you try it for at least a couple of days, since it might take a little bit of time to get used to.

The default animation length is perhaps a little bit too long ar the moment, but it can be configured to be shorter. But too short will look a bit choppy due to the fact that we don't get precise timings for the key presses. But I believe some smoothing can be added to fix that in a future release.

**Reply by 9mm** [reply]
Oh - everything is good on my end with the above settings. I actually do like having a few animations here and there, but subtle ones. For me its just perfect 🤩

**Reply by bodograumann** [reply]
Yeah, I guess using a lower duration, like 0.1, is definitely required.
The timing function is also not the best choice though. Looks like some ease-out, though I don't understand the implementation in detail.
Can this be made linear or faster at the end?

I'll create a bug report about the unintended scrolling then.
